### PR TITLE
Add heredoc syntax support for RUN, COPY, ADD (#245)

### LINKED ISSUE #245: RUN/COPY/ADD heredoc syntax not supported; treated as shell-form literal
## Summary

BuildKit supports heredoc syntax in `RUN`, `COPY`, and `ADD` instructions. The C# parser does not support heredoc syntax and treats `<<EOF` and subsequent lines as part of the shell-form command literal.

## Example

Input: `RUN <<EOF` followed by `echo hello` and `EOF` on separate lines.

C# output: The entire heredoc (including the `<<EOF` marker, the body, and the closing delimiter) is collapsed into one large `LiteralToken` as if it were shell-form command text.

Expected (BuildKit/Lean): Should produce a `heredoc` aggregate token with structured child tokens for the delimiter, body content lines, and closing delimiter.

## Affected Instructions

- `RUN <<DELIM`
- `COPY <<DELIM /path`
- `ADD <<DELIM /path`

## Impact

Any Dockerfile using heredoc syntax will be silently misparsed. The token structure will be wrong.

## Root Cause

The C# parser has no heredoc parser combinator. The `ParseHelper` does not recognize `<<` as a heredoc redirection operator. Adding heredoc support requires implementing a two-pass parser: detecting the opening `<<DELIM` during instruction parsing, then consuming the body until the closing delimiter.

### PR DESCRIPTION
**Pull request by mthalman**
## Summary

- Implements BuildKit-style heredoc parsing for RUN, COPY, and ADD instructions
- Split marker/body token architecture: `HeredocMarkerToken` inline in command stream, `HeredocBodyToken` sequential after command line
- `Heredoc` semantic wrapper pairs markers with bodies positionally
- New `HeredocDelimiterToken` (extends `IdentifierToken`) used consistently in both marker and body tokens
- Token decomposition follows existing codebase patterns (`SymbolToken` for `<<`, `-`, quotes)
- Supports multiple heredocs per instruction, chomp (`<<-`), quoted delimiters, empty bodies, CRLF
- COPY/ADD destination properly tokenized as separate `LiteralToken`
- `CommandInstruction.Command` now `virtual`; `RunInstruction` overrides as nullable for heredoc instructions
- `Heredocs` property (`IReadOnlyList<Heredoc>`) provides semantic access to heredoc content with chomp processing, delimiter name, and expand/quote metadata
- Round-trip fidelity preserved: `Parse(text).ToString() == text` for all syntax

Closes #245

## ⚠️ Breaking Changes

This release introduces nullability changes to three existing public properties. These changes are necessary to support heredoc instructions, which do not have the same token structure as traditional instructions (e.g., `RUN <<EOF` has no `Command` token; `COPY <<EOF /dest` has no separate destination token in the same position).

### 1. `CommandInstruction.Command`: `Command` → `Command?`

**Affected types:** `RunInstruction`, `CmdInstruction`, `EntrypointInstruction`, `ShellInstruction`

The `Command` property is now nullable. For heredoc `RUN` instructions, there is no `Command` token — the instruction body is expressed via heredoc bodies instead.

**Before:**
```csharp
RunInstruction run = dockerfile.Items.OfType<RunInstruction>().First();
string commandText = run.Command.ToString(); // always non-null
```

**After:**
```csharp
RunInstruction run = dockerfile.Items.OfType<RunInstruction>().First();
string? commandText = run.Command?.ToString(); // may be null for heredoc RUN
```

For non-heredoc instructions (`CMD`, `ENTRYPOINT`, `SHELL`, and traditional `RUN`), this property will continue to return a non-null value. The null case only occurs for heredoc `RUN` instructions. To access heredoc content, use the new `Heredocs` property:

```csharp
if (run.Command is null)
{
    // Heredoc RUN — access content via Heredocs
    foreach (Heredoc heredoc in run.Heredocs)
    {
        string body = heredoc.Content;    // chomp-processed body text
        string delimiter = heredoc.Name;  // e.g. "EOF"
        bool expand = heredoc.Expand;     // true if delimiter was not quoted
    }
}
```

### 2. `FileTransferInstruction.Destination`: `string` → `string?`

**Affected types:** `CopyInstruction`, `AddInstruction`

The `Destination` property is now nullable. For heredoc `COPY`/`ADD` instructions, the destination is still present in most cases, but the property accommodates edge cases where the token structure differs.

**Before:**
```csharp
CopyInstruction copy = dockerfile.Items.OfType<CopyInstruction>().First();
string dest = copy.Destination;
```

**After:**
```csharp
CopyInstruction copy = dockerfile.Items.OfType<CopyInstruction>().First();
string? dest = copy.Destination; // null-check if processing heredoc COPY/ADD
```

### 3. `FileTransferInstruction.DestinationToken`: `LiteralToken` → `LiteralToken?`

**Affected types:** `CopyInstruction`, `AddInstruction`

The underlying `DestinationToken` property is now nullable, consistent with the `Destination` change above.

**Before:**
```csharp
LiteralToken destToken = copy.DestinationToken;
```

**After:**
```csharp
LiteralToken? destToken = copy.DestinationToken;
```

### Migration guide

1. **If you only process non-heredoc Dockerfiles**, add null-forgiving operators or null checks to suppress warnings. The values will never actually be null for traditional instruction syntax:
   ```csharp
   string commandText = run.Command!.ToString();
   ```

2. **If you want to support heredoc Dockerfiles**, add null checks and use the new `Heredocs` property to access heredoc content:
   ```csharp
   if (run.Command is not null)
   {
       // Traditional RUN
       ProcessCommand(run.Command);
   }
   else
   {
       // Heredoc RUN
       foreach (Heredoc heredoc in run.Heredocs)
       {
           ProcessHeredocBody(heredoc.Content);
       }
   }
   ```

3. **If you set `Command` on a `RunInstruction`**, note that the setter now throws `InvalidOperationException` when called on a heredoc instruction. Check `Heredocs.Any()` first if the instruction might be a heredoc.

## Test plan

- [x] 1039 tests passing (0 failures)
- [x] Round-trip fidelity verified for single and multi-heredoc RUN, COPY, ADD
- [x] Token tree structure validated with `ValidateSymbol`, `ValidateIdentifier<HeredocDelimiterToken>` patterns
- [x] Edge cases: chomp, quoted delimiters, empty body, CRLF, special characters, multi-stage builds
- [x] All pre-existing tests unchanged and passing

### CODE REVIEW

File: .squad/agents/ripley/history.md [hunk 1]
```diff
@@ -95,3 +95,20 @@ Team update (2026-03-06T00:12:22Z): Phase 5 Capstone proofs completed: 12 new th
...
 ## Team update (2026-03-09T12:49:57Z): Copilot review workflow directive
 Added to shared decisions: Copilot PR review workflow (add reviewer via API, respond per comment, re-request until resolved). Note: `gh pr edit --add-reviewer` does NOT work for bots; must use API directly with bot name `copilot-pull-request-reviewer[bot]`. — decided by Scribe
+
+---
+
+## Team update (2026-03-10T12:30:00Z): Heredoc architecture design finalized and merged to decisions.md
+
+Designed comprehensive heredoc token architecture for Issue #245 (RUN, COPY, ADD instructions). Architecture decision merged to .squad/decisions.md.
+
+**Token Model:** `HeredocMarkerToken` (inline, contains `<<delimiter` with optional `<<<` chevron and EOF redirect) paired with `HeredocBodyToken` (sequential, contains body lines and closing delimiter). Semantic `Heredoc` wrapper encapsulates marker+body into semantic unit for instruction consumption.
```
[the remark this answers is not in the document]
Comment (mthalman) [reply .squad/agents/ripley/history.md]: Fixed in 9a5a295. Changed \ to \ (double chevron with dash for chomp mode) in Ripley's history entry.

File: src/Valleysoft.DockerfileModel/DockerfileParser.cs [hunk 2]
```diff
@@ -87,6 +208,18 @@ public static Dockerfile ParseContent(string text)
...
+    /// keyword that supports heredoc syntax (RUN, COPY, ADD).
+    /// </summary>
+    private static bool IsHeredocCapableInstruction(string constructText)
+    {
+        string trimmed = constructText.TrimStart();
+        return trimmed.StartsWith("RUN", StringComparison.OrdinalIgnoreCase)
+            || trimmed.StartsWith("COPY", StringComparison.OrdinalIgnoreCase)
+            || trimmed.StartsWith("ADD", StringComparison.OrdinalIgnoreCase);
+    }
+
```
[the remark this answers is not in the document]
Comment (mthalman) [reply src/Valleysoft.DockerfileModel/DockerfileParser.cs]: Fixed in 2f81ceb. Added a `StartsWithInstructionKeyword(text, keyword)` helper method that checks both the prefix match and a word boundary (next character must be whitespace or end-of-string). `IsHeredocCapableInstruction` now uses this helper, so "RUNNING" or "ADDITIONAL" no longer falsely match as heredoc-capable instructions.

File: src/Valleysoft.DockerfileModel/ParseHelper.cs [hunk 3]
```diff
@@ -968,6 +970,371 @@ private delegate Parser<IEnumerable<Token>> CreateWrappedTokenParserDelegate(
...
+        if (lineEndPos > cursor)
+        {
+            string restOfLine = source.Substring(cursor, lineEndPos - cursor);
+            string argsOnly = DockerfileParser.StripTrailingComment(restOfLine);
+            TokenizeRestOfLine(argsOnly, resultTokens, escapeChar);
+
+            // Preserve the comment portion as a StringToken for round-trip fidelity
+            if (argsOnly.Length < restOfLine.Length)
+            {
+                resultTokens.Add(new StringToken(restOfLine.Substring(argsOnly.Length)));
```
[the remark this answers is not in the document]
Comment (mthalman) [reply src/Valleysoft.DockerfileModel/ParseHelper.cs]: Already addressed — this line now uses `CommentToken.Parse(commentSegment)` instead of `new StringToken(...)`. The fix was made in commit 4f34e1c.

File: src/Valleysoft.DockerfileModel/ParseHelper.cs [hunk 4]
```diff
@@ -968,6 +970,407 @@ private delegate Parser<IEnumerable<Token>> CreateWrappedTokenParserDelegate(
...
+
+            int newLine = _line;
+            int newColumn = _column + 1;
+            if (_source[_position] == '\n')
+            {
+                newLine++;
+                newColumn = 1;
+            }
+            return new OffsetInput(_source, _position + 1, newLine, newColumn);
+        }
```
[the remark this answers is not in the document]
Comment (mthalman) [reply src/Valleysoft.DockerfileModel/ParseHelper.cs]: Fixed in 38643f4 — OffsetInput now passes through the existing Memos dictionary from the original input instead of allocating a fresh one. Both the AdvanceInput factory and Advance() method propagate the memo cache.
